src/atc/eh/EventHubCaptureExtractor.py

```diff
@@ -0,0 +1,262 @@
+            f.expr(
+                'regexp_extract_all(input_file_name(),"['
+                + self.partitioning
+                + ']=([0-9]+)/",1)'
+            ),
+        )
+        df = df.withColumn("y", f.element_at("_parts", 1))
+        df = df.withColumn("m", f.element_at("_parts", 2))
+        df = df.withColumn("d", f.element_at("_parts", 3))
+        if "h" in self.partitioning:
+            df = df.withColumn("h", f.element_at("_parts", 4))
+
+        df = df.withColumn(
+            "pdate",
+            f.expr(
+                "make_timestamp("
+                "cast(y AS INT),"
+                "cast(m AS INT),"
+                "cast(d AS INT),"
+                + ("cast(h AS INT)," if "h" in self.partitioning else "0,")
+                + '0,0,"UTC")'
+            ),
+        )
+        df = df.drop("_parts")
+        return df
+
+    def _now_utc(self):
+        return datetime.datetime.now(tz=datetime.timezone.utc)
+
+    def _end_edge_of_now_bin(self):
+        # the bin that covers the current time
+        return dt(
+            year=self._now_utc().year + 1,
+            month=1,
+            day=1,
+            hour=0,
+            second=0,
+            microsecond=0,
+            tzinfo=datetime.timezone.utc,
+        )
+
+    def _validated_slice_arguments(self, from_partition: dt, to_partition: dt):
+        from_partition = from_partition.astimezone(datetime.timezone.utc)
+
+        if not self._validate_timestamp(from_partition):
+            raise ValueError("from_partition is not a pure partition edge")
+
+        if to_partition is None:
+            # if the to_partition is not given, we set it to the end of
+            to_partition = self._end_edge_of_now_bin()
+        else:
+            # to_partition was given by the caller, we need to validate it.
+            to_partition = to_partition.astimezone(datetime.timezone.utc)
+
+            if not self._validate_timestamp(to_partition):
+                raise ValueError("to_partition is not a clean partition edge")
+
+            # ensure that limits are ordered correctly
+            if to_partition < from_partition:
+                from_partition, to_partition = to_partition, from_partition
+
+        if to_partition == from_partition:
+            raise ValueError("Requested to read over zero partitions.")
+
+        return from_partition, to_partition
+
+    def _break_into_partitioning_parts(self, fp: dt, tp: dt) -> List[str]:
+        """
+        data is grouped to use as few spark.read.load commands as possible
+        while still respecting the given limits exactly.
+        The slice is cut using the largest bite possible for each cut.
+        """
+        parts_to_load: List[str] = []
+
+        # We now have a slice description [fp, tp)
+        # we now take bites out of this slice until the slice is closed.
+        # in each iteration we take the biggest bite that we can,
+        while fp != tp and fp < self._now_utc():
```

Comment:
because it is implemented in each loop until it either becomes equal to tp or goes past the current time.
Say you want to read since last midnight. The first bite will be a day, then fp will be next midnight, still far from tp, but past the present time. No more bites are needed then.